cnxdb/migrations/20170911201035_transform_cnxml_to_html.py

```diff
@@ -0,0 +1,151 @@
+import logging
+import time
+from datetime import timedelta
+
+import rhaptos.cnxmlutils
+from dbmigrator import deferred, logger
+
+
+def _batcher(seq, size):
+    for pos in range(0, len(seq), size):
+         yield seq[pos:pos + size]
+
+
+def should_run(cursor, limit='LIMIT 1'):
+    version = rhaptos.cnxmlutils.__version__
+    version_text = '%data-cnxml-to-html-ver="{}"%'.format(version)
+    logger.info('Looking for {}'.format(version_text))
+    cursor.execute("""\
+WITH index_cnxml_html AS (
+    SELECT files.fileid, module_ident, file
+        FROM module_files NATURAL JOIN files
+        WHERE filename = 'index.cnxml.html'
+) SELECT fileid, max(module_ident)
+    FROM index_cnxml_html
+    WHERE convert_from(file, 'utf-8') NOT LIKE %s
+    GROUP BY fileid {}""".format(limit),
+                   (version_text,))
+    return cursor.fetchall()
+
+
+@deferred
+def up(cursor):
+    """Transform all index.cnxml to index.cnxml.html"""
+    # Get all the index.cnxml.html that does not have rhaptos.cnxmlutils
+    # current version
+    to_transform = should_run(cursor, limit='')
+    num_todo = len(to_transform)
+
+    batch_size = 10
+    logger.info('Batch size: {}'.format(batch_size))
+
+    start = time.time()
+    guesstimate = 1 * num_todo
+    guess_complete = guesstimate + start
+    logger.info('Completion guess: '
+                '"{}" ({})'.format(time.ctime(guess_complete),
+                                   timedelta(0, guesstimate)))
+
+    # Check if datamigrations.index_cnxml_html exists
+    cursor.execute("""\
+SELECT 1 FROM information_schema.tables
+    WHERE table_name = 'index_cnxml_html'
+      AND table_schema = 'datamigrations'""")
+    module_idents = tuple([i[1] for i in to_transform])
+    if not cursor.fetchall():
+        # Store module_files for modules we are going to update
+        cursor.execute("""\
+CREATE TABLE datamigrations.index_cnxml_html AS
+    SELECT * FROM module_files
+        WHERE module_ident IN %s
+          AND filename = 'index.cnxml.html'""",
+                       (module_idents,))
+    else:
+        cursor.execute("""\
+INSERT INTO datamigrations.index_cnxml_html
+    SELECT * FROM module_files
+        WHERE module_ident IN %s
+          AND filename = 'index.cnxml.html'
+          AND NOT EXISTS (
+            SELECT 1 FROM datamigrations.index_cnxml_html b
+                WHERE b.module_ident = module_files.module_ident);
+UPDATE datamigrations.index_cnxml_html b
+    SET fileid = module_files.fileid
+    FROM module_files
+    WHERE module_files.module_ident = b.module_ident
+      AND module_files.filename = b.filename
+      AND module_files.fileid != b.fileid
+      AND module_files.module_ident IN %s""",
+                       (module_idents, module_idents))
+
+    num_complete = 0
+    for batch in _batcher(to_transform, batch_size):
+        module_idents = tuple([i[1] for i in batch])
+        logger.info('Transform module_idents {}'.format(module_idents))
```

Comment:
This is noise during actual transform run -  `.debug` would be better, or remove